tests/frontend/test_io.py

```diff
@@ -320,18 +319,18 @@ def test_empty_program(self):
             io.to_program(bb)
 
     def test_gate_not_defined(self):
-        """Test empty program raises error"""
+        """Test unknown gate raises error"""
 
         bb_script = """\
         name test_program
         version 1.0
 
-        Toffoli | [0, 1]
+        np | [0, 1]
```

Comment:
How come you changed this?

Reply:
I realised there was a vulnerability with using
```python
try:
    getattr(ops, a_user_provided_string)(*args, **kwargs)
except Attribute error:
    raise Exception
```

in that 

1. Python turns _everything_ defined in the module as an accessible module attribute, i.e. if NumPy is imported, then `ops.np` exists.

2. The try/except pattern always causes the `getattr` line to be executed, and only fails if the attribute doesn't exist.

So, a malicious user would be able to do something like this:

```
np.io.something_malicious('rm -rf /') | [0, 1, 2]
```
and it could potentially be executed.

The new logic uses the `__all__` dictionary to verify the quantum operation is whitelisted, _before_ calling `getattr`.